GA4 Setup Variants

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #5275 

## Relevant technical choices

This PR updates the GA4 `<SetupBanner />` component to scaffold its variants according to the availability of the GA4 measurement ID and existing tag.

Difference from IB:
- > If `footer` is provided, render it within a new `Row` at the end of the `Grid`.
    - Displaying `footer` at the end of `<Grid />` becomes problematic when there are more elements in `<BannerNotification />` and it breaks in some cases as well. As a result, I decided to display it after the CTA actions within the same `<Cell />`.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

eugene-manuilov commented on assets/js/modules/analytics-4/components/dashboard/ActivationBanner/SetupBanner.js:
```diff
@@ -19,25 +19,98 @@
+	let ctaLabel;
+	let footer;
+
+	switch ( true ) {
+		case ! ga4MeasurementID && ! existingTag:
```
Please, use `if ( ... ) { ... } else if ( ... ) { ... } else if ( ... ) { ... } ...` approach instead of `switch( true )` magic.

eugene-manuilov commented on assets/js/components/notifications/BannerNotification/index.js:
```diff
@@ -396,27 +397,39 @@ function BannerNotification( {
-								</DismissComponent>
-							) }
+						{ footer && (
+							<p className="googlesitekit-publisher-win__footer-text">
+								{ footer }
+							</p>
```
I think if we go with this approach, then it should be a `<div>` tag instead of `<p>` because in the future we might want to use it somewhere else with a footer that consists of different block-level elements which can't be put within `<p>` tag.

That's said we need to update this to be something like `<div className="googlesitekit-publisher-win__footer">{ footer }</div>` and then in the `SetupBanner` we need to pass `<p className="...">...</p>` to the `footer` prop of the `BannerNotification` element. It also means that we need to move styles for the `.googlesitekit-publisher-win__footer-text` to be specific only for the `SetupBanner` component.

eugene-manuilov commented on assets/js/modules/analytics-4/components/dashboard/ActivationBanner/SetupBanner.js:
```diff
@@ -19,25 +19,98 @@
+						onCTAClick={ onCTAClick }
+						footer={ footer }
+						dismiss={ __( 'Cancel', 'google-site-kit' ) }
+					></BannerNotification>
```
```suggestion
					/>
```

eugene-manuilov commented on assets/js/modules/analytics-4/components/dashboard/ActivationBanner/index.stories.js:
```diff
@@ -16,15 +16,37 @@
+export const ReminderBanner = Template.bind( {} );
+ReminderBanner.storyName = 'ReminderBanner';
+
+export const SetupBanner = Template.bind( {} );
+SetupBanner.storyName = 'SetupBanner';
+SetupBanner.decorators = [
```
Let's create stories for each case that we can have based on measurement ID and existing tag combinations.

eugene-manuilov reviewed (changes requested):
Thanks, @nfmohit. This is a very good start. Mostly looks good to me, but added a few suggestions for you. Could you please take a look?

nfmohit commented:
> Thanks, @nfmohit. This is a very good start. Mostly looks good to me, but added a few suggestions for you. Could you please take a look?

Thank you very much for the very kind and valid review @eugene-manuilov! I have addressed them. Thank you!

eugene-manuilov commented on assets/js/modules/analytics-4/components/dashboard/ActivationBanner/SetupBanner.js:
```diff
@@ -19,25 +19,91 @@
+			'You can always add/edit this in the Site Kit Settings.',
+			'google-site-kit'
+		);
+	} else if ( ga4MeasurementID && ! existingTag ) {
+		title = __(
+			'Connect the Google Analytics 4 property that’s associated with your existing Universal Analytics property',
+			'google-site-kit'
+		);
+		ctaLabel = __( 'Connect', 'google-site-kit' );
+		footer = __(
+			'You can always add/edit this in the Site Kit Settings.',
+			'google-site-kit'
+		);
+	} else if ( ga4MeasurementID && existingTag ) {
+		title = __(
+			'Connect the Google Analytics 4 property that’s associated with your existing Universal Analytics property',
+			'google-site-kit'
+		);
+		ctaLabel = __( 'Connect', 'google-site-kit' );
+		footer = __(
+			'You can always add/edit this in the Site Kit Settings.',
+			'google-site-kit'
+		);
```
These two cases are identical, so we can combine them. I suggest using the following approach:


```js
if ( ga4MeasurementID ) {
    title = __( 'Connect the Google Analytics 4 property that’s associated with ...', ... );
    ctaLabel = __( ... );
    footer = __( ... );
} else if ( existingTag ) {
    title = __( ... );
    ctaLabel = __( ... );
    footer = __( 'A GA4 tag %s is found on this site but ...', ... );
} else {
    title = __( 'No existing Google Analytics 4 property found, ...', ... );
    ctaLabel = __( ... );
    footer = __( ... );
}
```

eugene-manuilov commented on assets/js/modules/analytics-4/components/dashboard/ActivationBanner/SetupBanner.stories.js:
```diff
@@ -0,0 +1,142 @@
+];
+
+export default {
+	title: 'Modules/Analytics4/ActivationBanner/SetupBanner',
```
Probably we should move the `SetupBanner` stories on the same level with the `ActivationBanner`?
```suggestion
	title: 'Modules/Analytics4/SetupBanner',
```

eugene-manuilov commented on assets/js/modules/analytics-4/components/dashboard/ActivationBanner/index.stories.js:
```diff
@@ -23,8 +23,8 @@ import ActivationBanner from './index';
 
-export const Default = Template.bind( {} );
-Default.storyName = 'Default';
+export const ReminderBanner = Template.bind( {} );
+ReminderBanner.storyName = 'ReminderBanner';
```
If we have just one story for the activation banner, then let's name it `ActivationBanner` and remove this piece from the title:

```
export const ActivationBanner = Template.bind( {} );
ActivationBanner.storyName = 'ActivationBanner';

export default {
    title: 'Modules/Analytics4',
    ...
}
```

eugene-manuilov commented on assets/js/modules/analytics-4/components/dashboard/ActivationBanner/SetupBanner.stories.js:
```diff
@@ -0,0 +1,142 @@
+					.receiveGetExistingTag( null );
+			};
+
+			setTimeout( async () => {
+				await waitForElement( () =>
+					document.querySelector(
+						'a.googlesitekit-notification__cta'
+					)
+				);
+				document
+					.querySelector( 'a.googlesitekit-notification__cta' )
+					.click();
+			}, 0 );
```
Why do we need this? Will the banner work without this?

nfmohit replied:
@eugene-manuilov Since the `<SetupBanner />` is displayed based on a local component state inside `<ActivationBanner />` and that state is only updated upon clicking on the CTA button on the banners, we need to simulate a click on the CTA to update the state inside `<ActivationBanner />` and display `<SetupBanner />`.

eugene-manuilov replied:
We don't need to use the `ActivationBanner` component here since it is intended for the setup banner only. Let's update the template to be as following and remove this timeout workaround:

```
import SetupBanner from './SetupBanner';

const Template = () => <SetupBanner ... />;
```

eugene-manuilov reviewed (changes requested):
Thanks, @nfmohit. Added a few more comments.

nfmohit commented:
> Thanks, @nfmohit. Added a few more comments.

Thank you for the kind review @eugene-manuilov! I have addressed your suggestions.

As for the Storybook stories, I restructured them based on our discussion in Slack, i.e. instead of showing a story for `ActivationBanner/index.stories.js`, we show individual stories for its child components, but nest them under `ActivationBanner` for context.

eugene-manuilov commented on assets/js/modules/analytics-4/components/dashboard/ActivationBanner/ReminderBanner.stories.js:
```diff
@@ -27,5 +27,5 @@ export const Default = Template.bind( {} );
 
 export default {
-	title: 'Modules/Analytics4/ActivationBanner',
+	title: 'Modules/Analytics4/ActivationBanner/ReminderBanner',
```
Let's use just `Modules/Analytics4` title here and rename `Default.storyName` to be `ReminderBanner`.

eugene-manuilov commented on assets/js/modules/analytics-4/components/dashboard/ActivationBanner/ReminderBanner.stories.js:
```diff
@@ -1,5 +1,5 @@
 /**
- * ActivationBanner Component Stories.
+ * ReminderBanner Component Stories.
```
Since this story is intended for the ReminderBanner component only, then we should use it instead of ActivationBanner in the template:

```
import ReminderBanner from './ReminderBanner';

const Template = () => <ReminderBanner ... />;
```

eugene-manuilov reviewed (changes requested):
Thanks, @nfmohit. We are almost ready to merge this. A few more comments regarding storybook stories and it should be good to go.

nfmohit commented:
Thank you for the additional corrections, @eugene-manuilov! I have addressed them.

eugene-manuilov reviewed (approved):
LGTM. Thanks, @nfmohit.
